contracts/l2/discovery/L2GNS.sol

```diff
@@ -0,0 +1,315 @@
+        );
+
+        ICuration curation = curation();
+
+        // Move all signal from previous version to new version
+        // NOTE: We will only do this as long as there is signal on the subgraph
+        if (subgraphData.nSignal != 0) {
+            // Burn all version signal in the name pool for tokens (w/no slippage protection)
+            // Sell all signal from the old deployment
+            uint256 tokens = curation.burn(
+                subgraphData.subgraphDeploymentID,
+                subgraphData.vSignal,
+                0
+            );
+
+            // Take the owner cut of the curation tax, add it to the total
+            // Upgrade is only callable by the owner, we assume then that msg.sender = owner
+            address subgraphOwner = msg.sender;
+            uint256 tokensWithTax = _chargeOwnerTax(
+                tokens,
+                subgraphOwner,
+                curation.curationTaxPercentage()
+            );
+
+            // Update pool: constant nSignal, vSignal can change (w/no slippage protection)
+            // Buy all signal from the new deployment
+            (subgraphData.vSignal, ) = curation.mint(_subgraphDeploymentID, tokensWithTax, 0);
+
+            emit SubgraphUpgraded(
+                _subgraphID,
+                subgraphData.vSignal,
+                tokensWithTax,
+                _subgraphDeploymentID
+            );
+        }
+
+        // Update target deployment
+        subgraphData.subgraphDeploymentID = _subgraphDeploymentID;
+
+        emit SubgraphVersionUpdated(_subgraphID, _subgraphDeploymentID, _versionMetadata);
+    }
+
+    /**
+     * @dev Receive a subgraph from L1.
+     * This function will initialize a subgraph received through the bridge,
+     * and store the migration data so that it's finalized later using finishSubgraphMigrationFromL1.
+     * @param _subgraphID Subgraph ID
+     * @param _subgraphOwner Owner of the subgraph
+     * @param _tokens Tokens to be deposited in the subgraph
+     * @param _nSignal Name signal for the subgraph in L1
+     */
+    function _receiveSubgraphFromL1(
+        uint256 _subgraphID,
+        address _subgraphOwner,
+        uint256 _tokens,
+        uint256 _nSignal
+    ) internal {
+        IL2GNS.SubgraphL2MigrationData storage migratedData = subgraphL2MigrationData[_subgraphID];
+        SubgraphData storage subgraphData = _getSubgraphData(_subgraphID);
+
+        subgraphData.reserveRatioDeprecated = fixedReserveRatio;
+        // The subgraph will be disabled until finishSubgraphMigrationFromL1 is called
+        subgraphData.disabled = true;
+        subgraphData.nSignal = _nSignal;
+
+        migratedData.tokens = _tokens;
+        migratedData.subgraphReceivedOnL2BlockNumber = block.number;
+
+        // Mint the NFT. Use the subgraphID as tokenID.
+        // This function will check the if tokenID already exists.
+        _mintNFT(_subgraphOwner, _subgraphID);
```

Comment:
Shouldn't the NFT mint happen on the `finishSubgraphMigrationFromL1` function? 

I can see how it could be done in both places but doing it here increases the callhook complexity (plus the NFT could be worthless if the migration is not finished).

Reply:
We do it here so that we can then use `onlySubgraphAuth` in `finishSubgraphMigrationFromL1`. You're right that it's not great to increase complexity in the callhook, but I thought this was still better than persisting the owner some other way and doing a custom ownership check

Reply:
Ah I missed that modifier. In that case I think adding a comment noting that would be helpful.

Reply:
Good point, I'll fix it in #764